### ISSUE TITLE
Bug, being triaged, may be OnionBalance or interaction with Tor

### ISSUE THREAD

**Issue by alecmuffett** [DonnchaC/onionbalance#62]
Symptoms, reported by 3rd party so it may take some time to get triage information:

* launch onionbalance for 2x master onions, which are served by 2x worker onions.
  * each worker serves both masters
* runs happily for several days
* onionbalance log suddenly reports (extended extract with context):

```
2017-04-15 17:00:58,029 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 17:10:57,706 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 17:20:58,387 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 17:21:59,896 [INFO]: Published a descriptor for service <MASTER1>.onion under replica 0.
2017-04-15 17:22:00,351 [INFO]: Published a descriptor for service <MASTER1>.onion under replica 1.
2017-04-15 17:22:00,800 [INFO]: Published a descriptor for service <MASTER2>.onion under replica 0.
2017-04-15 17:22:01,247 [INFO]: Published a descriptor for service <MASTER2>.onion under replica 1.
2017-04-15 17:30:57,857 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 17:31:04,726 [ERROR]: Received descriptor for instance <WORKER1>.onion with publication timestamp (2017-04-15 16:00:00) older than the latest descriptor (2017-04-15 17:00:
00). Ignoring the descriptor.
Exception in thread Event Notifier:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/stem/control.py", line 953, in _event_loop
    self._handle_event(event_message)
  File "/usr/local/lib/python2.7/dist-packages/stem/control.py", line 3727, in _handle_event
    listener(event_message)
  File "/usr/local/lib/python2.7/dist-packages/onionbalance/eventhandler.py", line 62, in new_desc_content
    descriptor.descriptor_received(descriptor_text)
  File "/usr/local/lib/python2.7/dist-packages/onionbalance/descriptor.py", line 261, in descriptor_received
    instance_changed |= instance.update_descriptor(parsed_descriptor)
TypeError: unsupported operand type(s) for |=: 'bool' and 'NoneType'

2017-04-15 17:40:57,537 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 17:50:58,218 [INFO]: Initiating fetch of descriptors for all service instances.
```
* From this point forwards, no further master descriptors are published
* Onionbalance logs claim stuff like the following:

```
2017-04-15 19:00:58,112 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 19:10:57,795 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 19:20:57,478 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 19:30:58,165 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 19:31:59,236 [INFO]: Our descriptor for instance <WORKER1>.onion was received too long ago (7254). The instance may be offline. Its introduction points will not be included in the master descriptor.
2017-04-15 19:31:59,236 [INFO]: Our descriptor for instance <WORKER2>.onion was received too long ago (7255). The instance may be offline. Its introduction points will not be included in the master descriptor.
2017-04-15 19:31:59,237 [WARNING]: Error generating descriptor: No introduction points for service <MASTER1>.onion.
2017-04-15 19:31:59,237 [WARNING]: Error generating descriptor: No introduction points for service <MASTER1>.onion.
2017-04-15 19:31:59,238 [INFO]: Our descriptor for instance <WORKER1>.onion was received too long ago (7854). The instance may be offline. Its introduction points will not be included in the master descriptor.
2017-04-15 19:31:59,238 [INFO]: Our descriptor for instance <WORKER2>.onion was received too long ago (7255). The instance may be offline. Its introduction points will not be included in the master descriptor.
2017-04-15 19:31:59,238 [WARNING]: Error generating descriptor: No introduction points for service <MASTER2>.onion.
2017-04-15 19:31:59,239 [WARNING]: Error generating descriptor: No introduction points for service <MASTER2>.onion.
2017-04-15 19:40:57,847 [INFO]: Initiating fetch of descriptors for all service instances.
2017-04-15 19:50:57,525 [INFO]: Initiating fetch of descriptors for all service instances.
```

...until the service eventually falls over due to stale descriptors.

Visual check of the associated Tor log (too big to scrub, sorry) shows no apparent issues.

### Is there a temporary fault condition causing OB to curl up and die? 

It continues to run, for several days, but reads no more fresh worker descriptors, and publishes no more fresh master descriptors.

**Comment by alecmuffett** [comment]
I can confirm this reproduces on:

* Raspbian 
* Tor 0.2.9.9
* OnionBalance 0.1.7

...with the same error message. 

I am going to upgrade Tor to 0.3.x latest/stable and try to repro again.

**Comment by DonnchaC** [comment]
Thanks for reporting this @alecmuffett. This should be fixed with #64, I'll make a release soon,
